src/NHibernate/Collection/Generic/SetHelpers/SetSnapShot.cs

```diff
@@ -114,12 +114,22 @@ IEnumerator IEnumerable.GetEnumerator()
 
 		void ICollection.CopyTo(Array array, int index)
 		{
-			if (!(array is T[] typedArray))
+			if (array is T[] typedArray)
 			{
-				throw new ArgumentException($"Array must be of type {typeof(T[])}.", nameof(array));
+				CopyTo(typedArray, index);
+				return;
 			}
 
-			CopyTo(typedArray, index);
+			if (array.GetType().GetElementType().IsAssignableFrom(typeof(T)))
```

Comment:
Do we need this check? Wouldn't it fail regardless? Do we need to test this behaviour?

Reply:
I don't think we actually need even first check. Implementation can simply be equal to generic CopyTo

Reply:
Ok. It can't be fully equal. What I meant:

```C#
void ICollection.CopyTo(Array array, int index)
{
	if (_hasNull)
		array.SetValue(default(T), index);
	((ICollection) _values.Keys).CopyTo(array, index + (_hasNull ? 1 : 0));
}
```